File: src/main/java/com/example/httpclientpropagation30/HttpClientPropagation30Application.java
package com.example.httpclientpropagation30;

import brave.baggage.BaggageField;
import brave.internal.baggage.ExtraBaggageContext;
import io.netty.channel.Channel;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.builder.SpringApplicationBuilder;
import org.springframework.cloud.sleuth.TraceContext;
import org.springframework.cloud.sleuth.brave.bridge.BraveTraceContext;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.client.reactive.ReactorClientHttpConnector;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.netty.ConnectionObserver;
import reactor.netty.NettyPipeline;
import reactor.netty.http.client.HttpClient;

import java.net.SocketAddress;
import java.util.*;
import java.util.function.Supplier;

@SpringBootApplication
class HttpClientPropagation30Application {

    public static void main(String[] args) {
        new SpringApplicationBuilder(HttpClientPropagation30Application.class).run(args);
    }
}

@Configuration
class HttpClientPropagation30Configuration {

    @Bean
    BaggageField valueToPropagate() {
        return BaggageField.create("some_field");
    }

    @Bean
    WebClient webClient(WebClient.Builder builder, HttpClient customHttpClient) {
        return builder.clientConnector(new ReactorClientHttpConnector(customHttpClient)).build();
    }

    @Bean
    HttpClient customHttpClient(BaggageField valueToPropagate) {
        return HttpClient.create()
                         .doOnChannelInit(
                                 (ConnectionObserver observer, Channel channel, SocketAddress remoteAddress) -> {
// -----> works here, but it's only called on first request
                                     System.out.println("baggage: " + valueToPropagate.getValue());

                                     var pipeline = channel.pipeline();
                                     if (pipeline.get(NettyPipeline.HttpCodec) != null) {
                                         Supplier<Map<String, String>> additionalContext = () -> createAdditionalContext(observer);

                                         var httpClientLoggingHandler = new MyNettyChannelHandler(additionalContext);
                                         pipeline.addAfter(NettyPipeline.HttpCodec, "MyHandler",
                                                           httpClientLoggingHandler);
                                     }
                                 });
    }

    private static Map<String, String> createAdditionalContext(ConnectionObserver observer) {
        var context = observer.currentContext();
        return context.<TraceContext>getOrEmpty(TraceContext.class)
                      .filter(BraveTraceContext.class::isInstance)
                      .map(traceContext -> ExtraBaggageContext.getAllValues(BraveTraceContext.toBrave(traceContext)))
                      .map(Map::copyOf)
                      .orElse(Collections.emptyMap());
    }
}

File: src/main/kotlin/com/example/httpclientpropagation30/Httpclientpropagation30Application.kt
package com.example.httpclientpropagation30

import io.micrometer.tracing.Baggage
import io.micrometer.tracing.Tracer
import io.netty.channel.*
import io.netty.handler.codec.http.*
import io.netty.util.internal.logging.*
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.client.reactive.ReactorClientHttpConnector
import org.springframework.web.bind.annotation.*
import org.springframework.web.reactive.function.client.ClientResponse
import org.springframework.web.reactive.function.client.WebClient
import reactor.core.publisher.Hooks
import reactor.core.publisher.Mono
import reactor.netty.*
import reactor.netty.http.client.HttpClient
import java.net.SocketAddress
import java.util.*
import java.util.function.Supplier

fun main(args: Array<String>) {
    Hooks.enableAutomaticContextPropagation()
    runApplication<Httpclientpropagation30Application>(*args)
}

@RestController
class MyController(private val webClient: WebClient, private val valueToPropagate: Baggage) {

    @RequestMapping("/first/{value}")
    fun first(@PathVariable value: String): Mono<String> {
        valueToPropagate.set(value).get()
        return webClient.get().uri("http://localhost:8080/remoteSystem").exchangeToMono { handleResponse(it) }
    }

    private fun handleResponse(response: ClientResponse) = when {
        response.statusCode().is2xxSuccessful -> response.bodyToMono(String::class.java)
        else -> Mono.error(RuntimeException())
    }

    @RequestMapping("/remoteSystem")
    fun remoteSystem(): Mono<String> = Mono.just("Some response")
}

@SpringBootApplication
@Configuration
class Httpclientpropagation30Application {

    @Bean
    fun valueToPropagate(tracer: Tracer) = tracer.createBaggage("some_field")!!

    @Bean
    fun customHttpClient(valueToPropagate: Baggage): HttpClient {
        return HttpClient.create()
                .doOnChannelInit { observer: ConnectionObserver, channel: Channel, remoteAddress: SocketAddress? ->
// -----> works here, but it's only called on first request
                    println("baggage: ${valueToPropagate.makeCurrent().get()}")

                    val pipeline = channel.pipeline()
                    if (pipeline[NettyPipeline.HttpCodec] != null) {
                        val additionalContext = Supplier<Map<String, String?>> {
                            mapOf(valueToPropagate.name() to valueToPropagate.makeCurrent().get())
                        }
                        val httpClientLoggingHandler = MyHandler(additionalContext)
                        pipeline.addAfter(NettyPipeline.HttpCodec, "MyHandler", httpClientLoggingHandler)
                    }
                }
    }

    @Bean
    fun webClient(builder: WebClient.Builder, customHttpClient: HttpClient) =
            builder.clientConnector(ReactorClientHttpConnector(customHttpClient)).build()
}

class MyHandler(private val additionalContext: Supplier<Map<String, String?>>) : ChannelDuplexHandler() {

    override fun channelRead(ctx: ChannelHandlerContext, msg: Any) {
// -----> here value is null
        println(additionalContext.get().toString())
        ctx.fireChannelRead(msg)
    }

    override fun write(ctx: ChannelHandlerContext, msg: Any, promise: ChannelPromise) {
// -----> here value is null
        println(additionalContext.get().toString())
        ctx.write(msg, promise)
    }
}
